### PR TITLE
Implement memoization

### PR DESCRIPTION
**Pull request by odersky**
Implement `memo(...)` function which caches its argument on first evaluation
and re-uses the cached value afterwards. The cache is placed next to the
method enclosing the memo(...) call.

`memo` is a member of package `compiletime`.

### CODE REVIEW

File: docs/docs/reference/metaprogramming/inline.md [hunk 1]
```diff
@@ -406,6 +406,72 @@ inline def fail(p1: => Any) = {
...
+like this:
+```
+  def f(y: Int) = costly(x) * y
+```
+the `costly(x)` subexpression would be recomputed each time `f` was called, even though
+its result is the same each time. With the addition of `memo(...)` the subexpression
+in the parentheses is computed only the first time and is cached for subsequent recalculuations.
+The memoized program expands to the following code:
+```
+class C(x: T) {
```
Comment (soronpo) [inline docs/docs/reference/metaprogramming/inline.md]: missing `[T]`

**Comment by soronpo** [comment]
(I need this now! :smile:) 
Since `memo` hides the caching mechanism, people may use it wrong (in a mutating/impure function) and get unexpected results:
```scala
class C[T](x: T) {
  var someState : Boolean = false //somewhere this is modified
  def costly(x : T) : Int = {
    if (someState) ???
    else ???
  }  
  def f(y: Int) = memo(costly(x)) * y
}
```

How do we prevent this? Is there a way for the compiler to verify the function is pure?
If not, I believe this should be documented as well.

**Comment by odersky** [comment]
@soronpo There is not currently a mechanism to indicate that a function is pure. There might be in the future, but it will take a while to get there. 

The problem you outline is exactly the same for a lazy val, so not specific to `memo`.

**Comment by odersky** [comment]
Superseded by #6967

**Pull request closed without merging**